libs/vkd3d/command.c

```diff
@@ -1587,7 +1587,36 @@ static ULONG STDMETHODCALLTYPE d3d12_command_allocator_Release(ID3D12CommandAllo
         vkd3d_free(allocator->framebuffers);
         vkd3d_free(allocator->passes);
 
-        /* All command buffers are implicitly freed when a pool is destroyed. */
+        if (vkd3d_config_flags & VKD3D_CONFIG_FLAG_RECYCLE_COMMAND_POOLS)
+        {
+            /* Don't want to do this unless we have to, so hide it behind a config.
+             * For well-behaving apps, we'll just bloat memory. */
+            if (pthread_mutex_lock(&device->mutex) == 0)
+            {
+                if (device->cached_command_allocator_count < ARRAY_SIZE(device->cached_command_allocators))
+                {
+                    /* Recycle the pool. Some games spam free/allocate pools,
+                     * even if it completely goes against the point of the API. */
+
+                    /* Have to free command buffers here if we're going to recycle,
+                     * otherwise DestroyCommandPool takes care of it. */
+                    VK_CALL(vkFreeCommandBuffers(device->vk_device, allocator->vk_command_pool,
+                            allocator->command_buffer_count, allocator->command_buffers));
```

Comment:
are we setting `àllocator->command_buffer_count` to 0 anywhere?

Reply:
This is in the path where allocator gets destroyed. command_buffer_count is set to 0 in command_allocator_Reset.

Reply:
We'd be double-freeing command buffers there though:
```c
    if (allocator->command_buffer_count)
    {
        VK_CALL(vkFreeCommandBuffers(device->vk_device, allocator->vk_command_pool,
                allocator->command_buffer_count, allocator->command_buffers));
        allocator->command_buffer_count = 0;
    }
```